File: packages/csharp/SourceRpc.Mqtt/Mqtt5Frame.cs
using System.Text;
using System.Text.Json;
using MessagePack;
using MessagePack.Resolvers;
using MQTTnet;
using MQTTnet.Protocol;

namespace SourceRpc.Mqtt;

/// <summary>
/// What a frame is called on MQTT 5, and how it is read back off it. The layout is
/// docs/mqtt5-frame-spec.md, and this is the C# half of what `Transports/Mqtt5Frame.ts` does.
///
/// **This is not the flat frame, and that is deliberate.** The connection-oriented transports carry
/// one self-addressing map because they have one link and nothing else to address with. MQTT has a
/// great deal else: the topic carries the addressee, `responseTopic` says where a reply goes,
/// `correlationData` pairs it, and `messageExpiryInterval` lets the broker drop a request its caller
/// has stopped waiting for. Putting a flat frame in the payload would throw all of that away, along
/// with the property the layout exists for - that a plain MQTT client with no msgrpc code can take
/// part, and that an operator can see why a call failed in MQTT Explorer without decoding anything.
///
/// What *is* shared is the model: both wire formats map to <see cref="RpcFrame"/>, so a call means
/// the same thing on either and only the spelling differs.
/// </summary>
public static class Mqtt5Frame
{
    /// <summary>Control properties are prefixed so a broker injecting its own cannot be mistaken for one.</summary>
    public const string Version = "mr-v";
    public const string Source = "mr-src";
    public const string Kind = "mr-kind";
    public const string Path = "mr-path";
    public const string Method = "mr-method";
    public const string Event = "mr-event";
    public const string Code = "mr-code";
    public const string ContractVersion = "mr-ver";
    public const string Ttl = "mr-ttl";
    public const string IdempotencyKey = "mr-idem";
    public const string Fence = "mr-fence";
    public const string Deferred = "mr-deferred";
    public const string Outcome = "mr-outcome";
    public const string Seq = "mr-seq";
    public const string Epoch = "mr-epoch";
    public const string Nonce = "mr-nonce";
    public const string Timestamp = "mr-ts";
    public const string Signature = "mr-sig";

    /// <summary>
    /// The frame version this speaks. 3 covers the owner fence, the deferred marker, the ticket
    /// outcome and the event cursor - all of them in the signature, for peers that sign.
    /// </summary>
    public const string FrameVersion = "3";

    /// <summary>Which per-peer topic a frame belongs on.</summary>
    public static string Channel(string kind) =>
        kind switch
        {
            "call" or "subscribe" or "unsubscribe" => "req",
            "result" or "error" or "ticket" => "rsp",
            _ => "evt"
        };

    /// <summary>Kinds that expect an answer, and so are the only ones entitled to say where it goes.</summary>
    public static bool IsRequest(string? kind) => kind is "call" or "subscribe" or "unsubscribe";

    /// <summary>
    /// A reply that ends the exchange, so the reply address held for it can be released.
    ///
    /// A result marked deferred is *not* final - it is the receipt, and the answer comes later as a
    /// ticket - and a ticket is final only on the outcome that ends it. Releasing on the receipt
    /// sends every later answer to a derived topic in this peer's own encoding, so a caller that
    /// named its own reply topic gets the receipt where it asked and the answer somewhere else.
    /// </summary>
    public static bool IsFinalReply(RpcFrame frame) =>
        frame.Kind switch
        {
            "ticket" => frame.Outcome != "progress",
            "result" => frame.Deferred != true,
            "error" => true,
            _ => false
        };

    /// <summary>`&lt;prefix&gt;/&lt;channel&gt;/&lt;peer&gt;` - the topic a frame of this kind for this peer belongs on.</summary>
    public static string TopicFor(string prefix, string channel, string peer) => $"{prefix}/{channel}/{peer}";

    /// <summary>The peer a topic addresses, or null when it is not one of ours.</summary>
    public static string? AddresseeOf(string prefix, string topic)
    {
        if (!topic.StartsWith(prefix + "/", StringComparison.Ordinal))
            return null;
        var rest = topic[(prefix.Length + 1)..];
        var slash = rest.IndexOf('/');
        return slash < 0 ? null : rest[(slash + 1)..];
    }

    /// <summary>Build the packet a frame travels in.</summary>
    public static MqttApplicationMessage ToPacket(RpcFrame frame, string topic, string? responseTopic, bool json, uint expirySeconds)
    {
        var builder = new MqttApplicationMessageBuilder()
            .WithTopic(topic)
            .WithPayload(EncodeBody(frame.Body, json))
            .WithContentType(json ? "application/json" : "application/msgpack")
            .WithPayloadFormatIndicator(json ? MqttPayloadFormatIndicator.CharacterData : MqttPayloadFormatIndicator.Unspecified)
            .WithUserProperty(Version, FrameVersion)
            .WithUserProperty(Source, frame.Src)
            .WithUserProperty(Kind, frame.Kind);

        if (frame.Corr is { Length: > 0 } corr)
            builder = builder.WithCorrelationData(Encoding.UTF8.GetBytes(corr));
        if (responseTopic is { Length: > 0 })
            // Only a request expects an answer, and only a request should expire. The expiry is the
            // caller's own remaining time, so the broker stops holding it the moment the caller
            // stops waiting - the two used to be set independently on the TypeScript side, and a
            // request outlived its caller's patience by twenty seconds by default.
            builder = builder.WithResponseTopic(responseTopic).WithMessageExpiryInterval(expirySeconds);

        builder = Add(builder, Path, frame.Path);
        builder = Add(builder, Method, frame.Method);
        builder = Add(builder, Event, frame.Event);
        builder = Add(builder, Code, frame.Code);
        builder = Add(builder, ContractVersion, frame.Ver);
        builder = Add(builder, Ttl, frame.Ttl?.ToString());
        builder = Add(builder, IdempotencyKey, frame.Idem);
        builder = Add(builder, Fence, frame.Fence);
        builder = Add(builder, Deferred, frame.Deferred == true ? "1" : null);
        builder = Add(builder, Outcome, frame.Outcome);
        builder = Add(builder, Seq, frame.Seq?.ToString());
        builder = Add(builder, Epoch, frame.Epoch);

        return builder.Build();
    }

    private static MqttApplicationMessageBuilder Add(MqttApplicationMessageBuilder builder, string name, string? value) =>
        string.IsNullOrEmpty(value) ? builder : builder.WithUserProperty(name, value);

    /// <summary>
    /// Read a frame's *addressing and control properties* off a packet, with no body.
    ///
    /// Split from the body deliberately, and it is a security boundary rather than tidiness: the
    /// payload is attacker-supplied bytes handed to a deserializer, and it must not be parsed until
    /// the frame carrying it has been shown to be authentic. Anything that can publish to a peer's
    /// request topic can otherwise reach the MessagePack reader with no key and no signature.
    ///
    /// A repeated control property is refused rather than resolved: MQTT permits one, and taking the
    /// first or the last would let a sender show one value to a check and another to the dispatcher.
    /// </summary>
    public static RpcFrame? Headers(MqttApplicationMessage packet, string addressee, out string? refusal)
    {
        refusal = null;
        var values = new Dictionary<string, string>(StringComparer.Ordinal);
        foreach (var property in packet.UserProperties ?? [])
        {
            if (!property.Name.StartsWith("mr-", StringComparison.Ordinal))
                continue;
            if (!values.TryAdd(property.Name, property.Value))
            {
                refusal = $"repeated control property {property.Name}";
                return null;
            }
        }

        if (!values.TryGetValue(Source, out var src) || string.IsNullOrEmpty(src))
        {
            refusal = "frame names no source";
            return null;
        }
        if (!values.TryGetValue(Kind, out var kind) || string.IsNullOrEmpty(kind))
        {
            refusal = "frame names no kind";
            return null;
        }

        var declared = packet.ContentType;
        if (!string.IsNullOrEmpty(declared) && declared != "application/msgpack" && declared != "application/json")
        {
            // An unknown content type used to fall back to msgpack, which is a guess about how to
            // read bytes somebody else chose - and the guess decides what the values mean.
            refusal = $"unknown content type '{declared}'";
            return null;
        }

        return new RpcFrame
        {
            V = 2,
            Src = src,
            Tgt = addressee,
            Kind = kind,
            Corr = packet.CorrelationData is { Length: > 0 } data ? Encoding.UTF8.GetString(data) : null,
            Path = Get(values, Path),
            Method = Get(values, Method),
            Event = Get(values, Event),
            Code = Get(values, Code),
            Ver = Get(values, ContractVersion),
            Ttl = long.TryParse(Get(values, Ttl), out var ttl) ? ttl : null,
            Idem = Get(values, IdempotencyKey),
            Fence = Get(values, Fence),
            // Presence, not value: `mr-deferred` is only ever sent as '1', and reading it as a
            // boolean would hydrate a ticket for a sender that wrote 'false'.
            Deferred = values.ContainsKey(Deferred) ? true : null,
            Outcome = Get(values, Outcome),
            Seq = long.TryParse(Get(values, Seq), out var seq) && seq >= 0 ? seq : null,
            Epoch = Get(values, Epoch)
        };
    }

    /// <summary>
    /// The same frame with its payload read - called only once the frame has been verified.
    /// </summary>
    public static RpcFrame? WithBody(RpcFrame frame, MqttApplicationMessage packet, out string? refusal)
    {
        refusal = null;
        try
        {
            return frame with { Body = DecodeBody(packet.PayloadSegment, packet.ContentType == "application/json") };
        }
        catch (Exception e)
        {
            refusal = $"undecodable payload: {e.Message}";
            return null;
        }
    }

    private static string? Get(Dictionary<string, string> values, string name) =>
        values.TryGetValue(name, out var value) && value.Length > 0 ? value : null;

    /// <summary>
    /// The payload is the body and nothing else - the argument array for a request, the value for a
    /// result - which is what makes a frame readable to a peer with no msgrpc code.
    /// </summary>
    public static byte[] EncodeBody(object? body, bool json)
    {
        if (body is null)
            return [];
        if (json)
            return JsonSerializer.SerializeToUtf8Bytes(body);
        // Serialized against the runtime type rather than `object`, so a result that is a record
        // writes as a map rather than failing in the primitive formatter.
        return MessagePackSerializer.Serialize(body.GetType(), body, ContractlessStandardResolver.Options);
    }

    /// <summary>
    /// Read a payload back, as whatever the encoding produced.
    ///
    /// Read under <see cref="MessagePackSecurity.UntrustedData"/>, because that is what it is. The
    /// standard options are documented as omitting all protections, including any bound on nesting
    /// depth - and `PrimitiveObjectFormatter` recurses once per nesting level, so a few kilobytes of
    /// repeated `0x91` is a StackOverflowException, which .NET cannot catch and which no `try`
    /// around this call can help with. The JSON reader needs no equivalent: its default MaxDepth of
    /// 64 throws cleanly.
    /// </summary>
    public static object? DecodeBody(ArraySegment<byte> payload, bool json)
    {
        if (payload.Count == 0)
            return null;
        if (json)
            return JsonSerializer.Deserialize<JsonElement>(payload.AsSpan());
        // As `object`, which gives primitives, object[] and Dictionary - the same shapes the
        // MessagePack hub protocol produces, so RpcFrame.Arg<T> reads them without knowing which
        // transport delivered them.
        return MessagePackSerializer.Deserialize<object?>(payload.ToArray(), UntrustedContractless);
    }

    /// <summary>Contractless, and with the depth and hash-collision guards that Standard omits.</summary>
    private static readonly MessagePackSerializerOptions UntrustedContractless =
        ContractlessStandardResolver.Options.WithSecurity(MessagePackSecurity.UntrustedData);
}

File: packages/rpc/src/Transports/Mqtt5Frame.ts
import { stringToUint8Array, uint8ArrayToString } from 'uint8array-extras'
import { Message, MessageType } from '../RPC/Core.js'
import {
    RpcCallInstanceMethodPayload,
    RpcErrorCode,
    RpcErrorPayload,
    RpcEventPayload,
    RpcMessage,
    RpcMessageType,
    RpcRemoteError,
    RpcSuccessPayload
} from '../RPC/Messages.js'

/**
 * Mapping between msgrpc messages and the MQTT 5 packet layout described in
 * docs/mqtt5-frame-spec.md.
 *
 * The point is that a peer needs no msgrpc code to take part: where to reply and how to correlate
 * come from the protocol's own Response Topic and Correlation Data, and everything else is a
 * readable user property. Kept separate from the transport so the mapping can be read, and tested,
 * without a broker.
 */

/** Control properties are prefixed so a broker or gateway injecting its own cannot be mistaken for one. */
export const MR = {
    version: 'mr-v',
    source: 'mr-src',
    kind: 'mr-kind',
    path: 'mr-path',
    method: 'mr-method',
    event: 'mr-event',
    code: 'mr-code',
    nonce: 'mr-nonce',
    timestamp: 'mr-ts',
    signature: 'mr-sig',
    contractVersion: 'mr-ver',
    /**
     * Milliseconds the caller will still wait, counted from when it sent. Carried alongside MQTT's
     * own messageExpiryInterval rather than instead of it: expiry is coarse (whole seconds), the
     * broker decrements it, and it stops at the broker - it says nothing about how long a frame then
     * sat in the receiving process. This is the caller's own statement, signed, and it survives
     * relaying through a transport that does not speak MQTT at all.
     */
    ttl: 'mr-ttl',
    /**
     * Names the command a request is an attempt at, when the caller distinguishes the two. Absent
     * means the correlation data is the name, so a redelivered packet is the same command and a
     * fresh attempt is a different one.
     */
    idempotencyKey: 'mr-idem',
    /**
     * The owner generation the caller observed for the instance it is addressing. Absent means an
     * unfenced call, which is the ordinary case.
     *
     * This travels or the fence does not exist. `RpcServerHandler.fenceRefusal` returns early when
     * the payload carries no fence, so a layout with no representation for one does not weaken the
     * check - it removes it, and a command whose instance was reassigned mid-flight runs under the
     * new owner with nothing said. Which is the failure a fence exists to prevent.
     */
    fence: 'mr-fence'
} as const

/**
 * Version 3 adds the owner fence to the signature. Version 2 covers contentType, the error code,
 * the declared contract version, the response topic, the ttl and the idempotency key; version 1
 * covered none of them, and a frame signed under one cannot verify under another. Bumped rather
 * than negotiated: a receiver that quietly accepted either would let an attacker choose the weaker,
 * and under version 2 the weaker choice is the one where stripping `mr-fence` costs nothing.
 */
export const FRAME_VERSION = '3'

/** Frame versions this build will accept. A frame announcing anything else is refused, not guessed at. */
export const SUPPORTED_FRAME_VERSIONS = new Set([FRAME_VERSION])

export type FrameKind = 'call' | 'subscribe' | 'unsubscribe' | 'result' | 'error' | 'event'

/** Which per-peer topic a frame belongs on. */
export type Channel = 'req' | 'rsp' | 'evt'

export interface OutboundFrame {
    kind: FrameKind
    channel: Channel
    /** The request id, carried as MQTT correlation data. Absent on events. */
    correlation?: string
    path?: string
    method?: string
    event?: string
    code?: string
    /** Contract version the caller declares, when it has one. */
    version?: string
    /** Milliseconds the caller will still wait. Drives the MQTT message expiry as well. */
    ttl?: number
    /** Names the command rather than this attempt at it, when the caller says so. */
    idempotencyKey?: string
    /**
     * The owner generation the caller observed, when it fences. Carried flat rather than as the
     * payload's `{ownerEpoch}` object, because a user property is a string and this layout has
     * nowhere to put an object.
     */
    fence?: string
    /** Encoded as the packet payload: arguments for a request, the value for a result. */
    body: unknown
}

const requestKind = (method: string): FrameKind =>
    method === 'on' ? 'subscribe' : method === 'off' || method === 'removeListener' ? 'unsubscribe' : 'call'

/** Kinds that expect an answer, and so are the only ones entitled to say where it should go. */
export const isRequestKind = (kind: string | undefined) => kind === 'call' || kind === 'subscribe' || kind === 'unsubscribe'

/** Undefined for anything this layout has no representation for, which the transport drops. */
export const toOutboundFrame = (message: Message): OutboundFrame | undefined => {
    const payload = message.payload as RpcMessage | undefined
    if (!payload) return undefined
    switch (payload.type) {
        case RpcMessageType.CallInstanceMethod: {
            const call = payload as RpcCallInstanceMethodPayload
            return {
                kind: requestKind(call.method),
                channel: 'req',
                correlation: call.id,
                path: call.path,
                method: call.method,
                version: call.version,
                ttl: call.ttl,
                idempotencyKey: call.idempotencyKey,
                fence: call.fence?.ownerEpoch,
                body: call.params
            }
        }
        case RpcMessageType.success: {
            const success = payload as RpcSuccessPayload
            return { kind: 'result', channel: 'rsp', correlation: success.id, body: success.result }
        }
        case RpcMessageType.error: {
            const error = payload as RpcErrorPayload
            return { kind: 'error', channel: 'rsp', correlation: error.id, code: error.code, body: error.error }
        }
        case RpcMessageType.event: {
            const event = payload as RpcEventPayload
            return { kind: 'event', channel: 'evt', event: event.event, path: event.path, body: event.params }
        }
        default:
            return undefined
    }
}

export interface InboundFrame {
    kind: string
    correlation?: string
    path?: string
    method?: string
    event?: string
    code?: string
    version?: string
    /** What the caller said it would still wait, already narrowed by anything the broker reported. */
    ttl?: number
    idempotencyKey?: string
    /** The owner generation the caller observed, when it fenced. See MR.fence. */
    fence?: string
    body: unknown
}

/** Undefined when the frame does not describe anything this RPC layer can dispatch. */
export const fromInboundFrame = (frame: InboundFrame): Message | undefined => {
    switch (frame.kind) {
        case 'call':
        case 'subscribe':
        case 'unsubscribe': {
            if (!frame.correlation || !frame.path || !frame.method) return undefined
            const payload: RpcCallInstanceMethodPayload = {
                type: RpcMessageType.CallInstanceMethod,
                id: frame.correlation,
                path: frame.path,
                method: frame.method,
                version: frame.version,
                ttl: frame.ttl,
                idempotencyKey: frame.idempotencyKey,
                // Absent rather than `{ownerEpoch: undefined}`: fenceRefusal tests the fence object
                // for presence, so an empty one would be a fence against nothing rather than none.
                ...(frame.fence ? { fence: { ownerEpoch: frame.fence } } : {}),
                // A caller that sends no payload means no arguments.
                params: Array.isArray(frame.body) ? frame.body : frame.body === undefined || frame.body === null ? [] : [frame.body]
            }
            return { type: MessageType.RequestMessage, payload }
        }
        case 'result': {
            if (!frame.correlation) return undefined
            const payload: RpcSuccessPayload = { type: RpcMessageType.success, id: frame.correlation, result: frame.body }
            return { type: MessageType.ResponseMessage, payload }
        }
        case 'error': {
            if (!frame.correlation) return undefined
            const payload: RpcErrorPayload = {
                type: RpcMessageType.error,
                id: frame.correlation,
                code: (frame.code ?? 'Exception') as RpcErrorCode,
                error: frame.body as RpcRemoteError | undefined
            }
            return { type: MessageType.ErrorMessage, payload }
        }
        case 'event': {
            if (!frame.event) return undefined
            const payload: RpcEventPayload = {
                type: RpcMessageType.event,
                event: frame.event,
                path: frame.path,
                params: Array.isArray(frame.body) ? frame.body : [frame.body]
            }
            return { type: MessageType.EventMessage, payload }
        }
        default:
            return undefined
    }
}

export type RawUserProperties = { [key: string]: string | string[] } | undefined

/**
 * Read the control properties, refusing any that appear more than once.
 *
 * MQTT permits a repeated user property, and mqtt.js surfaces repeats as an array. Taking the
 * first or the last would let a sender show one value to a check and a different one to the
 * dispatcher, so a repeat is an ambiguity to refuse rather than resolve.
 */
export const readControlProperties = (properties: RawUserProperties): { values: { [key: string]: string } } | { duplicate: string } => {
    const values: { [key: string]: string } = {}
    for (const [key, value] of Object.entries(properties ?? {})) {
        if (!key.startsWith('mr-')) continue
        if (Array.isArray(value)) return { duplicate: key }
        values[key] = value
    }
    return { values }
}

export const correlationToString = (correlation: Uint8Array | undefined) => (correlation ? uint8ArrayToString(correlation) : undefined)
export const correlationToBytes = (correlation: string | undefined) => (correlation ? stringToUint8Array(correlation) : undefined)
